README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# gmgm <img src="man/figures/logo.png" align="right" height="200" />

<!-- badges: start -->

[![CRAN
status](https://www.r-pkg.org/badges/version/gmgm)](https://cran.r-project.org/package=gmgm)
<!-- badges: end -->

Gaussian mixture graphical models include Bayesian networks and dynamic
Bayesian networks (their temporal extension) whose local probability
distributions are described by Gaussian mixture models. They are
powerful tools for graphically and quantitatively representing nonlinear
dependencies between continuous variables. The R package gmgm provides a
complete framework to create, manipulate, learn the structure and the
parameters, and perform inference in these models.

## Example

The following example illustrates how gmgm can be used to learn the
structure, the parameters and perform inference in a Gaussian mixture
Bayesian network.

Let’s take the NHANES dataset provided in this package, which includes
body composition data measured in 2148 adults aged 20 to 59 years in the
United States. 1848 observations are randomly chosen as the training
set, while the 300 remaining ones are chosen as the test set. In the
latter, one-third of the values of each column are randomly removed:

``` r
library(gmgm)
set.seed(0)

data(data_body)
obs_train <- sample.int(2148, 1848)
data_train <- data_body[obs_train, ]
data_test <- data_body[setdiff(1:2148, obs_train), ]

data_test_na <- data_test
data_test_na$GENDER[sample.int(300, 100)] <- NA
data_test_na$AGE[sample.int(300, 100)] <- NA
data_test_na$HEIGHT[sample.int(300, 100)] <- NA
data_test_na$WEIGHT[sample.int(300, 100)] <- NA
data_test_na$FAT[sample.int(300, 100)] <- NA
data_test_na$WAIST[sample.int(300, 100)] <- NA
data_test_na$GLYCO[sample.int(300, 100)] <- NA

print(data_test_na)

#> # A tibble: 300 x 8
#>       ID GENDER   AGE HEIGHT WEIGHT   FAT WAIST GLYCO
#>    <int>  <int> <int>  <dbl>  <dbl> <dbl> <dbl> <dbl>
#>  1 93731      0    NA    NA    NA    31.3 102.    4.9
#>  2 93758     NA    55   157.   75.8  NA   102.    6.5
#>  3 93761     NA    44   166.   80.2  31.8  NA     5.8
#>  4 93811      0    NA   164.   79.6  NA    90.9   5.3
#>  5 93814      1    41   169.   NA    39.1  90.7   4.6
#>  6 93824      1    52   159.   72.2  41.1  87.4   5.6
#>  7 93837      1    NA   167.  109.   48.3  NA    NA  
#>  8 93841      1    48   172.   96.5  NA    NA    NA  
#>  9 93849      1    40   165.   61.4  26    84.4   5.2
#> 10 93926     NA    31    NA    NA    36.9 130.    5.3
#> # ... with 290 more rows
```

At first, the Bayesian network is initialized with nodes corresponding
to the variables of the dataset and no arc:

``` r
gmbn_init <- add_nodes(NULL,
                       c("AGE", "FAT", "GENDER", "GLYCO", "HEIGHT", "WAIST",
                         "WEIGHT"))

network(gmbn_init)
```

<img src="man/figures/network_init.png"/>

Starting from this initial structure, the training set is used to select
the best subset of arcs (among predefined candidate arcs) and estimate
the parameters of each local Gaussian mixture model:

``` r
arcs_cand <- data.frame(from = c("AGE", "GENDER", "HEIGHT", "WEIGHT", NA, "AGE",
                                 "GENDER", "AGE", "FAT", "GENDER", "HEIGHT",
                                 "WEIGHT", "AGE", "GENDER", "HEIGHT"),
                        to = c("FAT", "FAT", "FAT", "FAT", "GLYCO", "HEIGHT",
                               "HEIGHT", "WAIST", "WAIST", "WAIST", "WAIST",
                               "WAIST", "WEIGHT", "WEIGHT", "WEIGHT"))
                                  
res_learn <- struct_learn(gmbn_init, data_train, arcs_cand = arcs_cand,
                          verbose = TRUE, max_comp = 3)

#> node AGE    bic_old = -1604165.22026861    bic_new = -6936.35703030545
#> node FAT    bic_old = -1101157.27026861    bic_new = -5240.27043715568
#> node GENDER    bic_old = -2188.72026861444    bic_new = 7570.83224030922
#> node GLYCO    bic_old = -32261.7452686144    bic_new = -1351.77102746833
#> node HEIGHT    bic_old = -25686922.7452686    bic_new = -6212.66741648412
#> node WAIST    bic_old = -8931942.60026862    bic_new = -5573.47194798866
#> node WEIGHT    bic_old = -6366683.95526861    bic_new = -7942.70714135656
```

The final structure contains 11 arcs:

``` r
network(res_learn$gmgm)
```

<img src="man/figures/network_learn.png"/>

If we focus for example on the variable WEIGHT, we can see that it
directly depends on the variable HEIGHT. In the related local Gaussian
mixture model, an optimal number of two mixture components has been
selected:

``` r
print(res_learn$gmgm$WEIGHT)

#> $alpha
#> [1] 0.6200512 0.3799488
#> 
#> $mu
#>             [,1]     [,2]
#> WEIGHT  70.78853  96.3321
#> HEIGHT 165.24341 168.4551
#> 
#> $sigma
#> $sigma[[1]]
#>           WEIGHT   HEIGHT
#> WEIGHT 169.32464 57.37758
#> HEIGHT  57.37758 84.23874
#> 
#> $sigma[[2]]
#>           WEIGHT   HEIGHT
#> WEIGHT 395.43744 64.55256
#> HEIGHT  64.55256 86.69264
#> 
#> 
#> attr(,"class")
#> [1] "gmm"
```

Graphically, we observe that these mixture components fit well with the
training data:

``` r
ellipses(res_learn$gmgm$WEIGHT, data_train)
```

<img src="man/figures/ellipses.png"/>

Now that the Bayesian network has been learned, it can be used to
perform inference on the test set, which allows to fill in the missing
values:

``` r
data_infer <- inference(res_learn$gmgm, data_test_na)

print(data_infer)

#> # A tibble: 300 x 7
#>      AGE   FAT  GENDER GLYCO HEIGHT WAIST WEIGHT
#>    <dbl> <dbl>   <dbl> <dbl>  <dbl> <dbl>  <dbl>
#>  1  31.7  31.3 0        4.9    173. 102.    88.4
#>  2  55    40.9 0.930    6.5    157. 102.    75.8
#>  3  44    31.8 0.145    5.8    166. 101.    80.2
#>  4  32.4  27.4 0        5.3    164.  90.9   79.6
#>  5  41.   39.1 1        4.6    169.  90.7   74.4
#>  6  52.   41.1 1        5.6    159.  87.4   72.2
#>  7  40.3  48.3 1        5.93   167. 119.   109. 
#>  8  48    42.1 1        6.08   172. 111.    96.5
#>  9  40    26.  1        5.2    165.  84.4   61.4
#> 10  31.   36.9 0.00688  5.3    174. 130.   128. 
#> # ... with 290 more rows
```

Focusing back on the variable WEIGHT, the values have been filled in
with a mean absolute percentage error (MAPE) of 10.2%:

``` r
pred_weight <- data_infer$WEIGHT[is.na(data_test_na$WEIGHT)]
actual_weight <- data_test$WEIGHT[is.na(data_test_na$WEIGHT)]

mape <- mean(abs(pred_weight - actual_weight) / actual_weight)

print(mape)

#> [1] 0.101956
```


NEWS.md
## gmgm 1.1.2

* Formatted the equations in the documentation to avoid HTML5 validation problems.

## gmgm 1.1.1

* Fixed formatting and typing errors in the documentation.
* Added slightly more details to the main documentation page.

## gmgm 1.1.0

* Created function `relevant`, which extracts the minimal sub-Gaussian mixture graphical model required to infer a subset of nodes.
* Changed the default values of several arguments to increase the speed of learning functions and make them easier to use:
  * `regul = NULL` → `regul = 0.01` (functions `AIC`, `BIC`, `logLik`, `em`)
  * `max_iter_em = Inf` → `max_iter_em = 100` (function `em`)
  * `max_iter_smem = Inf` → `max_iter_smem = 10` (function `smem`)
  * `max_iter_step = Inf` → `max_iter_step = 10` (function `stepwise`)
  * `max_iter_pem = Inf` → `max_iter_pem = 5` (functions `param_em`, `struct_em`)
  * `max_iter_sem = Inf` → `max_iter_sem = 5` (function `struct_em`)

## gmgm 1.0.2

* Replaced dplyr's scoped verbs (`_all`, `_at`, `_if`), which have been superseded by the use of function `across`.
* Modified function `gmm` to coerce the elements of the `gmm` object to type double.
* Provided a more detailed example in the README file.

## gmgm 1.0.1

* Removed a unit test of function `em` that failed with some BLAS/LAPACK implementations.

## gmgm 1.0.0

* Initial release.


R/gmm.R
#' Create a Gaussian mixture model
#'
#' This function creates a Gaussian mixture model as an object of S3 class
#' \code{gmm}. A Gaussian mixture model is a weighted sum of multivariate
#' Gaussian distributions:
#' \deqn{p(x) = \sum_{i = 1}^M \alpha_i \mathcal{N}(x | \mu_i, \Sigma_i)}
#' where \eqn{\alpha_i} is the \eqn{i}th mixture proportion such that
#' \eqn{\alpha_i > 0} and \eqn{\sum_{i = 1}^M \alpha_i = 1}, \eqn{\mu_i} the
#' mean vector and \eqn{\Sigma_i} the covariance matrix of the \eqn{i}th mixture
#' component (Bilmes, 1998). Since conditional distributions can be derived from
#' joint distributions, the \code{gmm} class is also used to work with
#' conditional Gaussian mixture models (see function \code{\link{conditional}}
#' to explicit their parameters). Finally, note that a one-component Gaussian
#' mixture model can be created with function \code{\link{add_var}} (by passing
#' \code{NULL} as argument \code{gmm}), which allows to quickly initialize a
#' \code{gmm} object that can be passed to a learning function.
#'
#' @param alpha A positive numeric vector containing the mixture proportions. If
#' the sum of these proportions is not 1, a normalization is performed by
#' dividing them by this sum.
#' @param mu A numeric matrix containing the mean vectors bound by column.
#' @param sigma A list containing the covariance matrices.
#' @param var A character vector containing the variable names (by default the
#' row names of \code{mu}).
#'
#' @return A list of class \code{gmm} containing the elements \code{alpha},
#' \code{mu} and \code{sigma} passed as arguments (completed with the variable
#' names passed as argument \code{var}).
#'
#' @references
#' Bilmes, J. A. (1998). A Gentle Tutorial of the EM Algorithm and its
#' Application to Parameter Estimation for Gaussian Mixture and Hidden Markov
#' Models. Technical report, International Computer Science Institute.
#'
#' @seealso \code{\link{gmbn}}, \code{\link{gmdbn}}
#'
#' @examples
#' gmm_1 <- gmm(alpha = c(0.2, 0.5, 0.3),
#'              mu = matrix(c(109, 91, 44, 160, 41, 99, 87, 27, 173, 40, 86, 65,
#'                            35, 161, 40),
#'                          nrow = 5),
#'              sigma = list(matrix(c(208, 240, 32, 17, -6, 240, 378, 40, 55, -38,
#'                                    32, 40, 15, -2, 1, 17, 55, -2, 47, -13, -6,
#'                                    -38, 1, -13, 127),
#'                                  nrow = 5),
#'                           matrix(c(242, 270, 82, 10, 49, 270, 363, 83, 44, 19,
#'                                    82, 83, 38, -2, 15, 10, 44, -2, 45, -7, 49,
#'                                    19, 15, -7, 137),
#'                                  nrow = 5),
#'                           matrix(c(109, 102, 41, 11, 29, 102, 128, 34, 38, 10,
#'                                    41, 34, 36, -9, 16, 11, 38, -9, 56, -5, 29,
#'                                    10, 16, -5, 138),
#'                                  nrow = 5)),
#'              var = c("WAIST", "WEIGHT", "FAT", "HEIGHT", "AGE"))
#'
#' @export

gmm <- function(alpha, mu, sigma, var = rownames(mu)) {
  if (!is.vector(alpha, "numeric")) {
    "alpha is not a numeric vector" %>%
      stop()
  }

  n_comp <- alpha %>%
    length()

  if (n_comp == 0) {
    "alpha is empty" %>%
      stop()
  }

  if (any(!is.finite(alpha))) {
    "alpha has non-finite elements" %>%
      stop()
  }

  if (any(alpha <= 0)) {
    "alpha has non-positive elements" %>%
      stop()
  }

  if (n_comp == 1) {
    alpha <- 1
  } else {
    alpha <- alpha %>%
      unname()
    sum_alpha <- alpha %>%
      sum()

    if (abs(sum_alpha - 1) > .Machine$double.eps) {
      alpha <- alpha / sum_alpha

      if (any(alpha <= 0)) {
        "alpha has non-positive elements" %>%
          stop()
      }
    }
  }

  if (!is.matrix(mu)) {
    "mu is not a numeric matrix" %>%
      stop()
  }

  n_var <- mu %>%
    nrow()

  if (n_var == 0) {
    "mu has no row" %>%
      stop()
  }

  if (ncol(mu) != n_comp) {
    "alpha and mu have incompatible sizes" %>%
      stop()
  }

  if (!is.numeric(mu)) {
    "mu is not numeric" %>%
      stop()
  }

  if (any(!is.finite(mu))) {
    "mu has non-finite elements" %>%
      stop()
  }

  if (!inherits(sigma, "list")) {
    "sigma is not a list of numeric matrices" %>%
      stop()
  }

  if (length(sigma) != n_comp) {
    "alpha and sigma have different lengths" %>%
      stop()
  }

  sigma %>%
    iwalk(function(sigma, comp) {
      if (!is.matrix(sigma)) {
        "sigma[[" %>%
          str_c(comp, "]] is not a numeric matrix") %>%
          stop()
      }

      if (nrow(sigma) != n_var | ncol(sigma) != n_var) {
        "mu and sigma[[" %>%
          str_c(comp, "]] have incompatible sizes") %>%
          stop()
      }

      if (!is.numeric(sigma)) {
        "sigma[[" %>%
          str_c(comp, "]] is not numeric") %>%
          stop()
      }

      if (any(!is.finite(sigma))) {
        "sigma[[" %>%
          str_c(comp, "]] has non-finite elements") %>%
          stop()
      }

      if (!isSymmetric(unname(sigma))) {
        "sigma[[" %>%
          str_c(comp, "]] is not symmetric") %>%
          stop()
      }

      if (inherits(tryCatch(chol(sigma), error = function(e) {return(e)}),
                   "error")) {
        "sigma[[" %>%
          str_c(comp, "]] is not positive definite") %>%
          stop()
      }
    })

  if (!is.vector(var, "character")) {
    "var is not a character vector" %>%
        stop()
  }

  if (length(var) != n_var) {
    "mu and var have incompatible sizes" %>%
      stop()
  }

  if (any(duplicated(var))) {
    "var has duplicated elements" %>%
      stop()
  }

  if (any(is.na(var))) {
    "var has NA elements" %>%
      stop()
  }

  if (any(!str_detect(var,
                      "^(\\.([A-Za-z_\\.]|$)|[A-Za-z])[A-Za-z0-9_\\.]*$"))) {
    "var contains invalid variable names" %>%
      stop()
  }

  mu <- mu %>%
    as.numeric() %>%
    matrix(n_var, dimnames = list(var))
  dimnames_sigma <- var %>%
    list(var)
  sigma <- sigma %>%
    map(function(sigma) {
      sigma %>%
        as.numeric() %>%
        matrix(n_var, dimnames = dimnames_sigma) %>%
        return()
    }) %>%
    unname()
  gmm <- list(alpha = alpha, mu = mu, sigma = sigma)
  class(gmm) <- "gmm"
  gmm %>%
    return()
}
